guides/recipes/anti-patterns.md

```diff
@@ -0,0 +1,843 @@
+    MyApp.Notifications.email_user(user_id, message)
+  end
+end
+
+defmodule MyApp.Workers.LongRunningReportGenerator do
+  use Oban.Worker,
+    queue: :reports,  # Limited concurrency queue
+    priority: 3       # Lower priority (higher number)
+
+  @impl Oban.Worker
+  def perform(%{args: %{"report_id" => report_id}}) do
+    MyApp.Reports.generate_report(report_id)
+  end
+end
+```
+
+### Additional Remarks
+
+Design your queue structure based on these job characteristics:
+
+| Characteristic | Considerations |
+|----------------|---------------|
+| **Duration** | Separate quick jobs from long-running ones |
+| **Resource usage** | Limit concurrency for resource-intensive operations |
+| **Priority** | Use both queues and priority values for fine-grained control |
+| **External dependencies** | Create dedicated queues for rate-limited external services |
+| **Business importance** | Ensure critical operations get processed first |
+
+The Smart Engine will dynamically adjust concurrency based on job execution patterns, but it works best when you start with thoughtfully designed queue structure.
+
+## Anti-Pattern 7: Insufficient Error Handling in Jobs
+
+### Problem
+
+Jobs with inadequate error handling lead to:
+
+- Unnecessary retries that waste resources
+- Masking of issues that should trigger alerts
+- Lack of context for troubleshooting
+- Unpredictable system behavior under failure conditions
+
+### Example
+
+```elixir
+defmodule MyApp.Workers.UserSynchronizer do
+  use Oban.Worker, queue: :sync, max_attempts: 20
+
+  @impl Oban.Worker
+  def perform(%{args: %{"user_id" => user_id}}) do
+    # Poor error handling - will retry even for expected errors
+    case MyApp.Users.fetch(user_id) do
+      {:ok, user} ->
+        MyApp.ExternalService.sync_user(user)
+
+      {:error, _reason} ->
+        # Generic error, will trigger retry without context
+        raise "Failed to sync user #{user_id}"
+    end
+  end
+end
+```
+
+### Refactoring
+
+Implement thoughtful error handling with appropriate retry behavior:
+
+```elixir
+defmodule MyApp.Workers.UserSynchronizer do
+  use Oban.Worker, queue: :sync, max_attempts: 5
+  require Logger
+
+  @impl Oban.Worker
+  def perform(%{args: %{"user_id" => user_id} = args, attempt: attempt}) do
+    case MyApp.Users.fetch(user_id) do
+      {:ok, user} ->
+        case MyApp.ExternalService.sync_user(user) do
+          {:ok, _} ->
+            :ok
+
+          {:error, %{status: 400, reason: reason}} ->
+            # Client error - don't retry
+            Logger.warn("User sync failed with client error: #{reason}", args: args)
+            {:discard, :client_error}
+
+          {:error, %{status: status}} when status >= 500 ->
+            # Server error - worth retrying with backoff
+            backoff = :math.pow(2, attempt) |> round()
+            Logger.warn("User sync failed with server error: #{status}", args: args)
+            {:snooze, backoff}
```

Comment:
I think this is a bad example on `:snooze`. I use it when some resource the job needs is not ready yet, but it should be.

Reply:
Although that may lead to another anti pattern? Infinitely snoozing jobs

Reply:
It's better to use the backoff provided by Oban itself, either through the worker (`backoff/1`) or the `Oban.Backoff` module. This would lead to infinitely snoozing jobs, which is bad. Also, I wouldn't include arbitrary logger warnings in example code.